### PR TITLE
#1257 notify projects about removed vacancies

### PR DESCRIPTION
**Pull request by krzyk**
#1257 
* implemented informing projects about removed vacancies
* updated test and added todo for slack testing

### CODE REVIEW

**Comment by g4s8** [comment]
@0crat in

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/zerocracy/farm/issues/1525#issuecomment-409990838))

@g4s8 Job [#1525](https://github.com/zerocracy/farm/issues/1525) is already in scope

**Comment by krzyk** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/zerocracy/farm/issues/1525#issuecomment-410016485))

@krzyk This is what I know about this job in [C3NDPUA8L](https://www.0crat.com/p/C3NDPUA8L), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#1525](https://github.com/zerocracy/farm/issues/1525) is [in scope](http://www.zerocracy.com/policy.html#14) for 9hr
  * The role is `REV`
  * The job is not assigned to anyone
  * The [budget](http://www.zerocracy.com/policy.html#4) is not set yet
  * Job [footprint](https://www.0crat.com/footprint/C3NDPUA8L?q=%7Bjob%3A%27gh%3Azerocracy%2Ffarm%231525%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
This pull request [#1525](https://github.com/zerocracy/farm/issues/1525) is assigned to @krzyk[/z](https://www.0crat.com/u/krzyk), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/2142110085); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by krzyk** [comment]
@g4s8 here also, my own PR and I'm the reviewer

**Comment by g4s8** [comment]
@0crat assign @carlosmiranda

**Comment by 0crat** [comment]
> @0crat assign @carlosmiranda  ([here](https://github.com/zerocracy/farm/issues/1525#issuecomment-410116261))

@g4s8 This pull request [#1525](https://github.com/zerocracy/farm/issues/1525) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/2150141098); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

File: src/main/groovy/com/zerocracy/stk/pmo/vacancies/remove_old_vacancies.groovy [hunk 1]
```diff
@@ -18,26 +18,35 @@ package com.zerocracy.stk.pmo.vacancies
...
+  int days = new Policy(farm).get('51.days', 32)
   new Vacancies(farm)
     .bootstrap()
-    .removeOlderThan(timestamp.minusDays(new Policy(farm).get('51.days', 32)))
+    .removeOlderThan(timestamp - Duration.ofDays(days)).each { pid ->
+      Project notify = farm.find("@id='${pid}'")[0]
+      claim.copy()
+        .type('Notify project')
+        .param(
+        'message',
```
Comment (carlosmiranda) [inline src/main/groovy/com/zerocracy/stk/pmo/vacancies/remove_old_vacancies.groovy]: @krzyk formatting here is off, we should indent this and succeeding 3 lines two more spaces.

File: src/main/groovy/com/zerocracy/stk/pmo/vacancies/remove_old_vacancies.groovy [hunk 2]
```diff
@@ -18,26 +18,35 @@ package com.zerocracy.stk.pmo.vacancies
...
+        new Par(
+          'Vacancy was removed because it expired after %d days (see §51)',
```
Comment (carlosmiranda) [inline src/main/groovy/com/zerocracy/stk/pmo/vacancies/remove_old_vacancies.groovy]: @krzyk Dangling comma at the end of this line, let's remove it.

File: src/main/groovy/com/zerocracy/stk/pmo/vacancies/remove_old_vacancies.groovy [hunk 3]
```diff
@@ -18,26 +18,35 @@ package com.zerocracy.stk.pmo.vacancies
...
+        ).say(days))
```
Comment (carlosmiranda) [inline src/main/groovy/com/zerocracy/stk/pmo/vacancies/remove_old_vacancies.groovy]: @krzyk let's move the final close-paren to the next line.

File: src/main/java/com/zerocracy/pmo/Vacancies.java [hunk 4]
```diff
@@ -153,22 +155,26 @@ public void remove(final Project project) throws IOException {
...
+     * @return List of project pids that had vacancies removed
      * @throws IOException If fails
      */
-    public void removeOlderThan(final ZonedDateTime date) throws IOException {
+    public Iterable<String> removeOlderThan(final Instant date)
+        throws IOException {
         try (final Item item = this.item()) {
+            final String xpath = new JoinedText(
+                "",
+                "/vacancies/vacancy[xs:dateTime(added) < ",
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/pmo/Vacancies.java]: @krzyk this is a bit hard to read. Let's use `FormattedText` to construct the xpath, and supress `LineLength` warning if necessary.

**Comment by krzyk** [comment]
@carlosmiranda updated, please take a look

File: src/main/java/com/zerocracy/pmo/Vacancies.java [hunk 5]
```diff
@@ -153,22 +156,23 @@ public void remove(final Project project) throws IOException {
...
+     * @return List of project pids that had vacancies removed
      * @throws IOException If fails
      */
-    public void removeOlderThan(final ZonedDateTime date) throws IOException {
+    public Iterable<String> removeOlderThan(final Instant date)
+        throws IOException {
         try (final Item item = this.item()) {
+            final String xpath = new FormattedText(
+                "/vacancies/vacancy[xs:dateTime(added) < xs:dateTime('%s')]",
+                date.toString()
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/pmo/Vacancies.java]: @krzyk we don't have to use `toString()` here

**Comment by krzyk** [comment]
@carlosmiranda updated

**Comment by g4s8** [comment]
@rultor merge

**Comment by krzyk** [comment]
@g4s8 is the issue with long pings causing lack of QA assignements in closed issues and PRs?

**Comment by g4s8** [comment]
@0crat out

**Comment by g4s8** [comment]
@krzyk looks like 0crat didn't receive `close` event.

**Comment by 0crat** [comment]
Job was finished in 32 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
> @0crat out ([here](https://github.com/zerocracy/farm/issues/1525#issuecomment-410439905))

@g4s8 The job [#1525](https://github.com/zerocracy/farm/issues/1525) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ypshenychka** [comment]
@carlosmiranda ping

**Comment by carlosmiranda** [comment]
@ypshenychka confirmed, thanks. Sorry for delay

**Comment by ypshenychka** [comment]
@carlosmiranda thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +20 point(s) just awarded to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**